app/services/bulk_ingest_service.rb

```diff
@@ -3,51 +3,136 @@ class BulkIngestService
   attr_reader :change_set_persister, :logger
   delegate :metadata_adapter, to: :change_set_persister
   delegate :query_service, to: :metadata_adapter
-  def initialize(change_set_persister:, logger: Valkyrie.logger)
-    @logger = logger
+  def initialize(change_set_persister:, klass: ScannedResource, logger: Valkyrie.logger)
     @change_set_persister = change_set_persister
+    @klass = klass
+    @logger = logger
   end
 
-  def attach_each_dir(base_directory:, property:, file_filter: nil)
+  def attach_each_dir(base_directory:, property: nil, file_filter: nil, **attributes)
     Dir["#{base_directory}/*"].sort.each do |subdir|
       next unless File.directory?(subdir)
       logger.info "Attaching #{subdir}"
-      attach_dir(base_directory: subdir, property: property, file_filter: file_filter)
+      attach_dir(base_directory: subdir, property: property, file_filter: file_filter, **attributes)
     end
   end
 
-  def attach_dir(base_directory:, property:, file_filter: nil)
+  # Attach files within a directory
+  # This may attach to existing resources (such as EphemeraFolder objects) using a property (e. g. "barcode")
+  # This may also create new resources (such as ScannedResource objects)
+  # @param base_directory [String] the path to the base directory
+  # @param property [String] the resource property (for an existing resource)
+  # @param file_filter [String] the filter used for matching against the filename extension
+  def attach_dir(base_directory:, property: nil, file_filter: nil, **attributes)
+    base_directory_path = Pathname.new(base_directory)
+
     property_value = File.basename(base_directory)
-    resource = property_query_service.find_by_string_property(property: property, value: property_value).first
+    if property && property_value
+      attach_dir_for_existing(directory_path: base_directory_path, property: property, value: property_value, file_filter: file_filter)
+    else
+      attach_dir_for_new(directory_path: base_directory_path, file_filter: file_filter, **attributes)
+    end
+  end
 
-    logger.info "Found #{resource.id} for #{property}:#{property_value}"
-    attach_files(base_directory: base_directory, resource: resource, file_filter: file_filter)
+  def attach_dir_for_existing(directory_path:, property:, value:, file_filter:)
+    root_resource = find_resource_by(property: property, value: value)
+    return unless root_resource
+    attach_files(path: directory_path, resource: root_resource, file_filter: file_filter)
```

Comment:
I agree that this looks like it could be refactored to be simpler.  One possibility would be using `attach_children` here.  We currently only have a single-volume use case for attaching ephemera images to existing objects, but I can imagine using a similar workflow for multi-volume items if they were scanned by a vendor.

If that was the case, then you could just have the logic to find an existing resource or fallback on creating a new resource.